x-pack/filebeat/input/cel/input.go

```diff
@@ -0,0 +1,909 @@
+			// Additional fields may be present at the root of the object.
+			// The evaluation is repeated with the new state, after removing
+			// the events field, if the "want_more" field is present and true
+			// and a non-zero events array is returned.
+			//
+			// If cursor is present it must be either a single object or an
+			// array with the same length as events; each element i of the
+			// cursor array will be the details for obtaining the events at or
+			// beyond event i in events. If the cursor is a single object it
+			// is will be the details for obtaining events after the last
+			// event in the events array and will only be retained on
+			// successful publication of all the events in the array.
+			//
+			// If rate_limit is present it should be a map with numeric fields
+			// rate and burst. It may also have a string error field and
+			// other fields which will be logged. If it has an error field
+			// the rate and burst will not be used to set rate limit behaviour.
+			//
+			// The status code and rate_limit values may be omitted if they do
+			// not contribute to control.
+			//
+			// The following details how a cursor array works:
+			//
+			// Result after request resulting in 5 events. Each c obtained with
+			// an e points to the ~next e.
+			//
+			//    +----+   +----+        +----+
+			//    | e1 |   | c1 |        | e1 |
+			//    +----+   +----+        +----+   +----+
+			//    | e2 |   | c2 |        | e2 | < | c1 |
+			//    +----+   +----+        +----+   +----+
+			//    | e3 |   | c3 |        | e3 | < | c2 |
+			//    +----+   +----+   =>   +----+   +----+
+			//    | e4 |   | c4 |        | e4 | < | c3 |
+			//    +----+   +----+        +----+   +----+
+			//    | e5 |   | c5 |        | e5 | < | c4 |
+			//    +----+   +----+        +----+   +----+
+			//                           |next| < | c5 |
+			//                           +----+   +----+
+			//
+			// After a successful publication this will leave a single c and
+			// and empty events array. So the next evaluation has a boot.
+			//
+			// If the publication fails or execution is terminated at some
+			// point during the events array, we may end up with, e.g.
+			//
+			//    +----+  +----+        +----+   +----+
+			//    | e3 |  | c3 |        |next| < | c3 |
+			//    +----+  +----+        +----+   +----+
+			//    | e4 |  | c4 |   =>
+			//    +----+  +----+          lost events
+			//    | e5 |  | c5 |
+			//    +----+  +----+
+			//
+			// At this point, the c3 cursor (or at worst the c2 cursor) has
+			// been stored and we can continue from that point, recovering
+			// the lost events and potentially re-requesting e3.
+
+			var ok bool
+			ok, waitUntil, err = handleResponse(log, state, limiter)
+			if err != nil {
+				return err
+			}
+			if !ok {
+				continue
+			}
+
+			_, ok = state["url"]
+			if !ok && goodURL != "" {
+				state["url"] = goodURL
+				log.Infow("adding missing url", "state", mapstr.M(state), "url", goodURL)
```

Comment:
I think it would be helpful to clarify how it's correcting the issue and explain why it's happening. I'm thinking  something like

> adding the current URL to the returned state because it did not include a 'url'

I expect some use cases that only hit a single static URL will omit adding the `url` from their CEL expression. Should we make this debug level?

Reply:
Semantically, it feels like INFO, but in the case that people don't return a URL, it would spam the logs with this, though that would be easily fixed by following the documentation.